src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -931,7 +931,7 @@ private void AddAllocatedEndpointInfo(IEnumerable<AppResource> resources)
 
                 sp.EndpointAnnotation.AllocatedEndpoint = new AllocatedEndpoint(
                     sp.EndpointAnnotation,
-                    sp.EndpointAnnotation.IsProxied ? svc.AllocatedAddress! : "localhost",
+                    "localhost",
```

Comment:
The reason why `AllocatedAddress` is used here is because if one requests `localhost:1234`, but 1234 is already taken on one of the addresses behind `localhost` name (dual-stack machine), DCP will bind to specific IP address (IPv4 or IPv6) to satisfy the requested port. It will then return the specific IP address that was used.

If the client uses `localhost` to connect in that case, this will still work, but may be quite slow as the client tries the "wrong" address several times before switching to the other one.

I am not sure if you need to change this line as part of this change

Reply:
Will do

Reply:
This doesn't work because we use this allocated endpoint address to communicate with these endpoints. It seems like dcp returns 0.0.0.0 instead of localhost, 127.0.0.1 or ::1.

Reply:
Returns 0.0.0.0 for what kind of data in EndpointAnnotation?

Reply:
It's less about what it returns and more about how what is returned gets used. This is the "client" side of the service and we hand this out for other APIs to connect to. Much like port = client port and target port = listen port. It seems like we need that split for address as well.

Reply:
Do we need changes or new tests in https://github.com/dotnet/aspire/blob/main/tests/Aspire.Hosting.Tests/Dcp/ApplicationExecutorTests.cs?

Reply:
> It's less about what it returns and more about how what is returned gets used. This is the "client" side of the service and we hand this out for other APIs to connect to. Much like port = client port and target port = listen port. It seems like we need that split for address as well.

Not sure I follow. The localhost gets special treatment, but nothing else does. Meaning, if you tell DCP something is listening on address other than localhost (including 0.0.0.0 pseudo-address), it will use that address for the proxy, and this is the address the clients should be using too. So I am not sure why we need to force localhost here.

Reply:
Chatted with @davidfowl and understand now, 0.0.0.0, as-is, is not useable by clients. Will need to get that fixed on DCP side.